gctree/isotype.py

```diff
@@ -0,0 +1,432 @@
+#! /usr/bin/env python
+# -*- coding: utf-8 -*-
+
+import re
+import random
+import ete3
+from gctree.utils import hamming_distance
+import pickle
+import argparse
+from pathlib import Path
+import warnings
+from typing import Dict, List
+
+
+class IsotypeTemplate:
+    def __init__(self, isotype_order):
+        self.order = isotype_order
+
+    def new(self, isotype_name):
+        return Isotype(self.order, isotype_name)
+
+
+class Isotype:
+    # From https://doi.org/10.7554/eLife.16578.012, for humans:
+    # order = ["M", "G3", "A1", "G2", "G4", "E", "A2"]
+
+    def __init__(self, order: List[str], isotype_name: str):
+        self.order = order
+        if isotype_name == "?":
+            self.isotype = None
+        else:
+            try:
+                self.isotype = self.order.index(isotype_name)
+            except ValueError as e:
+                raise ValueError(
+                    "Unrecognized isotype name. Check that default"
+                    " or provided isotype name list contains all"
+                    " observed isotypes\n" + str(e)
+                )
+
+    def isbefore(self, t: "Isotype") -> bool:
+        return self.isotype <= t.isotype
+
+    def __repr__(self) -> str:
+        return f"Isotype('{self.order[self.isotype]}')"
+
+    def __str__(self) -> str:
+        return f"{self.order[self.isotype]}"
+
+    def copy(self) -> "Isotype":
+        return Isotype(self.order, self.order[self.isotype])
+
+    def __eq__(self, other: "Isotype") -> bool:
+        return self.isotype == other.isotype
+
+    def __hash__(self) -> int:
+        return hash(self.__repr__())
+
+    def resolutions(self) -> List["Isotype"]:
+        """Returns list of all possible isotypes if passed an ambiguous
+        isotype.
+
+        Otherwise, returns a list containing a copy of the passed
+        isotype.
+        """
+        if self.isotype is None:
+            return [Isotype(self.order, name) for name in self.order]
+        else:
+            return [self.copy()]
+
+
+def isotype_distance(t1: Isotype, t2: Isotype) -> float:
+    """This function is not symmetric on its arguments."""
+    if not t1.isbefore(t2):
+        return float("inf")
+    else:
+        return float(t1.isotype != t2.isotype)
```

Comment:
This will return 0 or 1, but do we want to count more than 1 switching event? E.g. should the distance between "M" and "A1" be 2?

Reply:
I talked about this with Erick and his feeling was no I think? But I'll ask in the thread. Based on the DAG in this paper `https://doi.org/10.7554/eLife.16578.012` it doesn't seem that simple, but maybe I should allow some sort of user-supplied transition weights. that should be easy